TCC模式下，TC db在二阶段宕机，一段时间后恢复，未提交的全局事务无法重试

- [x] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
TCC模式下，TM向TC发送GlobalCommitRequest，TC开始提交各分支事务。
在提交过程中，**TC端的db服务器宕机**，数据库连接抛出异常，异常在doGlobalCommit内被捕获后执行 globalSession.queueToRetryCommit()。
因数据库连接异常，无法将全局事务的状态更改为 CommitRetrying。因此，即使数据库在一段时间后恢复，该未提交的全局事务也无法被重试，并且以Committing的状态留在数据库中，导致最终不一致。

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
00:21:22.356 ERROR --- [rverHandlerThread_1_9_500] i.s.c.e.AbstractExceptionHandler         : Catch TransactionException while do RPC, request: xid=192.168.0.17:8091:110005065393115136,extraData=null
==>
io.seata.core.exception.TransactionException: global commit request failed. xid=192.168.0.17:8091:110005065393115136, msg=wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 50
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:98)
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:90)
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:116)
	at io.seata.server.AbstractTCInboundHandler.handle(AbstractTCInboundHandler.java:90)
	at io.seata.core.protocol.transaction.GlobalCommitRequest.handle(GlobalCommitRequest.java:34)
	at io.seata.server.coordinator.DefaultCoordinator.onRequest(DefaultCoordinator.java:431)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.onRequestMessage(ServerOnRequestProcessor.java:116)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.process(ServerOnRequestProcessor.java:77)
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:278)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.seata.common.exception.StoreException: wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 50
	at io.seata.server.storage.db.store.LogStoreDataBaseDAO.updateGlobalTransactionDO(LogStoreDataBaseDAO.java:232)
	at io.seata.server.storage.db.store.DataBaseTransactionStoreManager.writeSession(DataBaseTransactionStoreManager.java:104)
	at io.seata.server.storage.db.session.DataBaseSessionManager.addGlobalSession(DataBaseSessionManager.java:88)
	at io.seata.server.session.GlobalSession.queueToRetryCommit(GlobalSession.java:657)
	at io.seata.server.coordinator.DefaultCore.doGlobalCommit(DefaultCore.java:236)
	at io.seata.server.coordinator.DefaultCore.commit(DefaultCore.java:166)
	at io.seata.server.coordinator.DefaultCoordinator.doGlobalCommit(DefaultCoordinator.java:167)
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:95)
	... 12 common frames omitted
Caused by: com.alibaba.druid.pool.GetConnectionTimeoutException: wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 50
	at com.alibaba.druid.pool.DruidDataSource.getConnectionInternal(DruidDataSource.java:1732)
	at com.alibaba.druid.pool.DruidDataSource.getConnectionDirect(DruidDataSource.java:1404)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1384)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1374)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:98)
	at io.seata.server.storage.db.store.LogStoreDataBaseDAO.updateGlobalTransactionDO(LogStoreDataBaseDAO.java:225)
	... 19 common frames omitted
Caused by: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure

......

00:21:32.262 ERROR --- [verHandlerThread_1_11_500] i.s.c.e.AbstractExceptionHandler         : Catch TransactionException while do RPC, request: xid=192.168.0.17:8091:110005065393115136,extraData=null
==>
io.seata.core.exception.TransactionException: global commit request failed. xid=192.168.0.17:8091:110005065393115136, msg=wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 70
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:98)
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:90)
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:116)
	at io.seata.server.AbstractTCInboundHandler.handle(AbstractTCInboundHandler.java:90)
	at io.seata.core.protocol.transaction.GlobalCommitRequest.handle(GlobalCommitRequest.java:34)
	at io.seata.server.coordinator.DefaultCoordinator.onRequest(DefaultCoordinator.java:431)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.onRequestMessage(ServerOnRequestProcessor.java:116)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.process(ServerOnRequestProcessor.java:77)
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:278)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.seata.common.exception.DataAccessException: wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 70
	at io.seata.server.storage.db.store.LogStoreDataBaseDAO.queryGlobalTransactionDO(LogStoreDataBaseDAO.java:127)
	at io.seata.server.storage.db.store.DataBaseTransactionStoreManager.readSession(DataBaseTransactionStoreManager.java:157)
	at io.seata.server.storage.db.session.DataBaseSessionManager.findGlobalSession(DataBaseSessionManager.java:162)
	at io.seata.server.session.SessionHolder.findGlobalSession(SessionHolder.java:312)
	at io.seata.server.session.SessionHolder.findGlobalSession(SessionHolder.java:301)
	at io.seata.server.coordinator.DefaultCore.commit(DefaultCore.java:143)
	at io.seata.server.coordinator.DefaultCoordinator.doGlobalCommit(DefaultCoordinator.java:167)
	at io.seata.server.AbstractTCInboundHandler$2.execute(AbstractTCInboundHandler.java:95)
	... 12 common frames omitted
Caused by: com.alibaba.druid.pool.GetConnectionTimeoutException: wait millis 5004, active 0, maxActive 100, creating 0, createErrorCount 70
	at com.alibaba.druid.pool.DruidDataSource.getConnectionInternal(DruidDataSource.java:1732)
	at com.alibaba.druid.pool.DruidDataSource.getConnectionDirect(DruidDataSource.java:1404)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1384)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1374)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:98)
	at io.seata.server.storage.db.store.LogStoreDataBaseDAO.queryGlobalTransactionDO(LogStoreDataBaseDAO.java:116)
	... 19 common frames omitted
Caused by: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure

```


### Ⅲ. Describe what you expected to happen
数据库恢复后，未提交的全局事务可以被重试。

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

    @GlobalTransactional
    public void business() {
        FeignClientA.try();
        FeignClientB.try();
    }

1. 在二阶段TC提交分支事务A的过程中，关闭TC端的数据库服务
2. 等到TM输出 commit status: Committing 日志后，重启TC端的数据库服务
3. 此后会发现该未提交的全局事务无法被重试

### Ⅴ. Anything else we need to know?
TM的部分日志：
```
......

2021-03-02 00:20:51.455  INFO 20006 --- [nio-8084-exec-1] i.seata.tm.api.DefaultGlobalTransaction  : Begin new global transaction [192.168.0.17:8091:110005065393115136]
2021-03-02 00:21:22.249 ERROR 20006 --- [nio-8084-exec-1] i.s.c.r.n.AbstractNettyRemotingClient    : wait response error:cost 30001 ms,ip:127.0.0.1:8091,request:xid=192.168.0.17:8091:110005065393115136,extraData=null
2021-03-02 00:21:22.250 ERROR 20006 --- [nio-8084-exec-1] i.seata.tm.api.DefaultGlobalTransaction  : Failed to report global commit [192.168.0.17:8091:110005065393115136],Retry Countdown: 5, reason: RPC timeout
2021-03-02 00:21:27.265  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : channel [id: 0xc6cc5fbc, L:/127.0.0.1:50076 - R:/127.0.0.1:8091] read idle.
2021-03-02 00:21:27.265  INFO 20006 --- [ctor_TMROLE_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : will destroy channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 - R:/127.0.0.1:8091]
2021-03-02 00:21:27.265  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 - R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:27.268  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:27.268  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.netty.NettyClientChannelManager  : return to pool, rm channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : channel valid false,channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : will destroy channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : channel inactive: [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : channel valid false,channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : will destroy channel:[id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:27.269  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.n.AbstractNettyRemotingClient    : ChannelHandlerContext(AbstractNettyRemotingClient$ClientHandler#0, [id: 0xc6cc5fbc, L:/127.0.0.1:50076 ! R:/127.0.0.1:8091]) will closed
2021-03-02 00:21:31.431  INFO 20006 --- [eoutChecker_1_1] i.s.c.r.netty.NettyClientChannelManager  : will connect to 127.0.0.1:8091
2021-03-02 00:21:31.431  INFO 20006 --- [eoutChecker_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : NettyPool create channel to transactionRole:TMROLE,address:127.0.0.1:8091,msg:< RegisterTMRequest{applicationId='business-service', transactionServiceGroup='business-service-fescar-service-group'} >
2021-03-02 00:21:31.434  INFO 20006 --- [eoutChecker_1_1] i.s.c.rpc.netty.TmNettyRemotingClient    : register TM success. client version:1.4.0, server version:1.4.1,channel:[id: 0x258a4a90, L:/127.0.0.1:50165 - R:/127.0.0.1:8091]
2021-03-02 00:21:31.434  INFO 20006 --- [eoutChecker_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : register success, cost 2 ms, version:1.4.1,role:TMROLE,channel:[id: 0x258a4a90, L:/127.0.0.1:50165 - R:/127.0.0.1:8091]
2021-03-02 00:21:32.363  INFO 20006 --- [ctor_TMROLE_1_1] i.s.c.r.p.c.ClientOnResponseProcessor    : msg: 10 is not found in futures.
2021-03-02 00:21:42.267  INFO 20006 --- [nio-8084-exec-1] i.seata.tm.api.DefaultGlobalTransaction  : [192.168.0.17:8091:110005065393115136] commit status: Committing
```


### Ⅵ. Environment:

- JDK version : 1.8
- OS : MacOS
- Seata version: 1.4.1
- Store mode: db
- RPC: Feign

--- a/changes/1.5.0.md
+++ b/changes/1.5.0.md
@@ -43,6 +43,8 @@ Seata 是一款开源的分布式事务解决方案，提供高性能和简单�
   - [[#3481](https://github.com/seata/seata/pull/3481)] 修复当 consul client 获取集群信息报错时会导致刷新任务中断
   - [[#3491](https://github.com/seata/seata/pull/3491)] 修复README.md文件中的拼写错误
   - [[#3531](https://github.com/seata/seata/pull/3531)] 修复RedisTransactionStoreManager读取brachTransaction中的NPE
+  - [[#3500](https://github.com/seata/seata/pull/3500)] 修复 oracle 和 postgreSql 不能查询出 column info 的问题
+  - [[#3560](https://github.com/seata/seata/pull/3560)] 修复 Committing 状态的事务没有时间阈值问题
 
   ### optimize： 
 
@@ -103,6 +105,7 @@ Seata 是一款开源的分布式事务解决方案，提供高性能和简单�
   - [github-ganyu](https://github.com/github-ganyu)
   - [xuande](https://github.com/xuande)
   - [tanggen](https://github.com/tanggen)
+  - [dmego](https://github.com/dmego)
 
 
 同时，我们收到了社区反馈的很多有价值的issue和建议，非常感谢大家。

--- a/changes/en-us/1.5.0.md
+++ b/changes/en-us/1.5.0.md
@@ -44,6 +44,8 @@
   - [[#3481](https://github.com/seata/seata/pull/3481)] fix seata node refresh failure because consul crash
   - [[#3491](https://github.com/seata/seata/pull/3491)] fix typo in README.md
   - [[#3531](https://github.com/seata/seata/pull/3531)] fix the NPE of RedisTransactionStoreManager when get branch transactions
+  - [[#3500](https://github.com/seata/seata/pull/3500)] fix oracle and postgreSQL can't query column info
+  - [[#3560](https://github.com/seata/seata/pull/3560)] set time threshold for transactions in committing state
 
 
   ### optimize： 	
@@ -108,6 +110,7 @@
   - [github-ganyu](https://github.com/github-ganyu)
   - [xuande](https://github.com/xuande)
   - [tanggen](https://github.com/tanggen)
+  - [dmego](https://github.com/dmego)
 
 
   Also, we receive many valuable issues, questions and advices from our community. Thanks for you all.	

--- a/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
+++ b/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
@@ -281,7 +281,7 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
         SessionHelper.forEach(rollbackingSessions, rollbackingSession -> {
             try {
                 // prevent repeated rollback
-                if (rollbackingSession.getStatus().equals(GlobalStatus.Rollbacking) && !rollbackingSession.isRollbackingDead()) {
+                if (rollbackingSession.getStatus().equals(GlobalStatus.Rollbacking) && !rollbackingSession.isDeadSession()) {
                     //The function of this 'return' is 'continue'.
                     return;
                 }
@@ -316,6 +316,11 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
         long now = System.currentTimeMillis();
         SessionHelper.forEach(committingSessions, committingSession -> {
             try {
+                // prevent repeated commit
+                if (committingSession.getStatus().equals(GlobalStatus.Committing) && !committingSession.isDeadSession()) {
+                    //The function of this 'return' is 'continue'.
+                    return;
+                }
                 if (isRetryTimeout(now, MAX_COMMIT_RETRY_TIMEOUT.toMillis(), committingSession.getBeginTime())) {
                     /**
                      * Prevent thread safety issues

--- a/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
+++ b/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
@@ -17,9 +17,7 @@ package io.seata.server.coordinator;
 
 import java.time.Duration;
 import java.util.Collection;
-import java.util.HashSet;
 import java.util.Map;
-import java.util.Set;
 import java.util.concurrent.ScheduledThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 import io.netty.channel.Channel;
@@ -275,12 +273,12 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle retry rollbacking.
      */
     protected void handleRetryRollbacking() {
-        Set<GlobalSession> retryRollbackingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.RollbackRetrying);
-        if (CollectionUtils.isEmpty(retryRollbackingSet)) {
+        Collection<GlobalSession> rollbackingSessions = SessionHolder.getRetryRollbackingSessionManager().allSessions();
+        if (CollectionUtils.isEmpty(rollbackingSessions)) {
             return;
         }
         long now = System.currentTimeMillis();
-        SessionHelper.forEach(retryRollbackingSet, rollbackingSession -> {
+        SessionHelper.forEach(rollbackingSessions, rollbackingSession -> {
             try {
                 // prevent repeated rollback
                 if (rollbackingSession.getStatus().equals(GlobalStatus.Rollbacking) && !rollbackingSession.isRollbackingDead()) {
@@ -300,10 +298,7 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 rollbackingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                boolean commitResult = core.doGlobalRollback(rollbackingSession, true);
-                if (commitResult) {
-                    GlobalSession.reliableSessionMap.remove(GlobalStatus.RollbackRetrying);
-                }
+                core.doGlobalRollback(rollbackingSession, true);
             } catch (TransactionException ex) {
                 LOGGER.info("Failed to retry rollbacking [{}] {} {}", rollbackingSession.getXid(), ex.getCode(), ex.getMessage());
             }
@@ -314,12 +309,12 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle retry committing.
      */
     protected void handleRetryCommitting() {
-        Set<GlobalSession> retryCommittingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.CommitRetrying);
-        if (CollectionUtils.isEmpty(retryCommittingSet)) {
+        Collection<GlobalSession> committingSessions = SessionHolder.getRetryCommittingSessionManager().allSessions();
+        if (CollectionUtils.isEmpty(committingSessions)) {
             return;
         }
         long now = System.currentTimeMillis();
-        SessionHelper.forEach(retryCommittingSet, committingSession -> {
+        SessionHelper.forEach(committingSessions, committingSession -> {
             try {
                 if (isRetryTimeout(now, MAX_COMMIT_RETRY_TIMEOUT.toMillis(), committingSession.getBeginTime())) {
                     /**
@@ -331,10 +326,7 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 committingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                boolean commitResult = core.doGlobalCommit(committingSession, true);
-                if (commitResult) {
-                    GlobalSession.reliableSessionMap.remove(GlobalStatus.CommitRetrying);
-                }
+                core.doGlobalCommit(committingSession, true);
             } catch (TransactionException ex) {
                 LOGGER.info("Failed to retry committing [{}] {} {}", committingSession.getXid(), ex.getCode(), ex.getMessage());
             }
@@ -349,11 +341,12 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle async committing.
      */
     protected void handleAsyncCommitting() {
-        Set<GlobalSession> asyncCommittingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.AsyncCommitting);
-        if (CollectionUtils.isEmpty(asyncCommittingSet)) {
+        Collection<GlobalSession> asyncCommittingSessions = SessionHolder.getAsyncCommittingSessionManager()
+                .allSessions();
+        if (CollectionUtils.isEmpty(asyncCommittingSessions)) {
             return;
         }
-        SessionHelper.forEach(asyncCommittingSet, asyncCommittingSession -> {
+        SessionHelper.forEach(asyncCommittingSessions, asyncCommittingSession -> {
             try {
                 // Instruction reordering in DefaultCore#asyncCommit may cause this situation
                 if (GlobalStatus.AsyncCommitting != asyncCommittingSession.getStatus()) {
@@ -361,40 +354,11 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 asyncCommittingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                boolean commitResult = core.doGlobalCommit(asyncCommittingSession, true);
-                if (commitResult) {
-                    GlobalSession.reliableSessionMap.remove(GlobalStatus.AsyncCommitting);
-                }
+                core.doGlobalCommit(asyncCommittingSession, true);
             } catch (TransactionException ex) {
                 LOGGER.error("Failed to async committing [{}] {} {}", asyncCommittingSession.getXid(), ex.getCode(), ex.getMessage(), ex);
             }
         });
-    }
-
-    /**
-     * Get global session from store and reliableMap
-     * @param status global session status
-     * @return session status set
-     */
-    private Set<GlobalSession> getGlobalSessionFormStoreAndReliableMap(GlobalStatus status) {
-        Set<GlobalSession> sessionSet = new HashSet<>();
-        Set<GlobalSession> reliableSessions = GlobalSession.reliableSessionMap.get(status);
-        Collection<GlobalSession> storeSessions = null;
-        if (GlobalStatus.CommitRetrying == status) {
-            storeSessions  = SessionHolder.getRetryCommittingSessionManager().allSessions();
-        } else if (GlobalStatus.AsyncCommitting == status) {
-            storeSessions  = SessionHolder.getAsyncCommittingSessionManager().allSessions();
-        } else if (GlobalStatus.RollbackRetrying == status) {
-            storeSessions  = SessionHolder.getRetryRollbackingSessionManager().allSessions();
-        }
-
-        if (CollectionUtils.isNotEmpty(storeSessions)) {
-            sessionSet.addAll(storeSessions);
-        }
-        if (CollectionUtils.isNotEmpty(reliableSessions)) {
-            sessionSet.addAll(reliableSessions);
-        }
-        return sessionSet;
     }
 
     /**

--- a/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
+++ b/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
@@ -17,7 +17,9 @@ package io.seata.server.coordinator;
 
 import java.time.Duration;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.ScheduledThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 import io.netty.channel.Channel;
@@ -273,12 +275,12 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle retry rollbacking.
      */
     protected void handleRetryRollbacking() {
-        Collection<GlobalSession> rollbackingSessions = SessionHolder.getRetryRollbackingSessionManager().allSessions();
-        if (CollectionUtils.isEmpty(rollbackingSessions)) {
+        Set<GlobalSession> retryRollbackingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.RollbackRetrying);
+        if (CollectionUtils.isEmpty(retryRollbackingSet)) {
             return;
         }
         long now = System.currentTimeMillis();
-        SessionHelper.forEach(rollbackingSessions, rollbackingSession -> {
+        SessionHelper.forEach(retryRollbackingSet, rollbackingSession -> {
             try {
                 // prevent repeated rollback
                 if (rollbackingSession.getStatus().equals(GlobalStatus.Rollbacking) && !rollbackingSession.isRollbackingDead()) {
@@ -298,7 +300,10 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 rollbackingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                core.doGlobalRollback(rollbackingSession, true);
+                boolean commitResult = core.doGlobalRollback(rollbackingSession, true);
+                if (commitResult) {
+                    GlobalSession.reliableSessionMap.remove(GlobalStatus.RollbackRetrying);
+                }
             } catch (TransactionException ex) {
                 LOGGER.info("Failed to retry rollbacking [{}] {} {}", rollbackingSession.getXid(), ex.getCode(), ex.getMessage());
             }
@@ -309,12 +314,12 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle retry committing.
      */
     protected void handleRetryCommitting() {
-        Collection<GlobalSession> committingSessions = SessionHolder.getRetryCommittingSessionManager().allSessions();
-        if (CollectionUtils.isEmpty(committingSessions)) {
+        Set<GlobalSession> retryCommittingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.CommitRetrying);
+        if (CollectionUtils.isEmpty(retryCommittingSet)) {
             return;
         }
         long now = System.currentTimeMillis();
-        SessionHelper.forEach(committingSessions, committingSession -> {
+        SessionHelper.forEach(retryCommittingSet, committingSession -> {
             try {
                 if (isRetryTimeout(now, MAX_COMMIT_RETRY_TIMEOUT.toMillis(), committingSession.getBeginTime())) {
                     /**
@@ -326,7 +331,10 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 committingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                core.doGlobalCommit(committingSession, true);
+                boolean commitResult = core.doGlobalCommit(committingSession, true);
+                if (commitResult) {
+                    GlobalSession.reliableSessionMap.remove(GlobalStatus.CommitRetrying);
+                }
             } catch (TransactionException ex) {
                 LOGGER.info("Failed to retry committing [{}] {} {}", committingSession.getXid(), ex.getCode(), ex.getMessage());
             }
@@ -341,12 +349,11 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      * Handle async committing.
      */
     protected void handleAsyncCommitting() {
-        Collection<GlobalSession> asyncCommittingSessions = SessionHolder.getAsyncCommittingSessionManager()
-            .allSessions();
-        if (CollectionUtils.isEmpty(asyncCommittingSessions)) {
+        Set<GlobalSession> asyncCommittingSet = getGlobalSessionFormStoreAndReliableMap(GlobalStatus.AsyncCommitting);
+        if (CollectionUtils.isEmpty(asyncCommittingSet)) {
             return;
         }
-        SessionHelper.forEach(asyncCommittingSessions, asyncCommittingSession -> {
+        SessionHelper.forEach(asyncCommittingSet, asyncCommittingSession -> {
             try {
                 // Instruction reordering in DefaultCore#asyncCommit may cause this situation
                 if (GlobalStatus.AsyncCommitting != asyncCommittingSession.getStatus()) {
@@ -354,11 +361,40 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
                     return;
                 }
                 asyncCommittingSession.addSessionLifecycleListener(SessionHolder.getRootSessionManager());
-                core.doGlobalCommit(asyncCommittingSession, true);
+                boolean commitResult = core.doGlobalCommit(asyncCommittingSession, true);
+                if (commitResult) {
+                    GlobalSession.reliableSessionMap.remove(GlobalStatus.AsyncCommitting);
+                }
             } catch (TransactionException ex) {
                 LOGGER.error("Failed to async committing [{}] {} {}", asyncCommittingSession.getXid(), ex.getCode(), ex.getMessage(), ex);
             }
         });
+    }
+
+    /**
+     * Get global session from store and reliableMap
+     * @param status global session status
+     * @return session status set
+     */
+    private Set<GlobalSession> getGlobalSessionFormStoreAndReliableMap(GlobalStatus status) {
+        Set<GlobalSession> sessionSet = new HashSet<>();
+        Set<GlobalSession> reliableSessions = GlobalSession.reliableSessionMap.get(status);
+        Collection<GlobalSession> storeSessions = null;
+        if (GlobalStatus.CommitRetrying == status) {
+            storeSessions  = SessionHolder.getRetryCommittingSessionManager().allSessions();
+        } else if (GlobalStatus.AsyncCommitting == status) {
+            storeSessions  = SessionHolder.getAsyncCommittingSessionManager().allSessions();
+        } else if (GlobalStatus.RollbackRetrying == status) {
+            storeSessions  = SessionHolder.getRetryRollbackingSessionManager().allSessions();
+        }
+
+        if (CollectionUtils.isNotEmpty(storeSessions)) {
+            sessionSet.addAll(storeSessions);
+        }
+        if (CollectionUtils.isNotEmpty(reliableSessions)) {
+            sessionSet.addAll(reliableSessions);
+        }
+        return sessionSet;
     }
 
     /**

--- a/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
+++ b/server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java
@@ -347,7 +347,7 @@ public class DefaultCoordinator extends AbstractTCInboundHandler implements Tran
      */
     protected void handleAsyncCommitting() {
         Collection<GlobalSession> asyncCommittingSessions = SessionHolder.getAsyncCommittingSessionManager()
-                .allSessions();
+            .allSessions();
         if (CollectionUtils.isEmpty(asyncCommittingSessions)) {
             return;
         }

--- a/server/src/main/java/io/seata/server/session/GlobalSession.java
+++ b/server/src/main/java/io/seata/server/session/GlobalSession.java
@@ -157,10 +157,10 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
     }
 
     /**
-     * prevent could not handle rollbacking transaction
-     * @return if true force roll back
+     * prevent could not handle committing and rollbacking transaction
+     * @return if true retry commit or roll back
      */
-    public boolean isRollbackingDead() {
+    public boolean isDeadSession() {
         return (System.currentTimeMillis() - beginTime) > (2 * 6000);
     }
 
@@ -687,5 +687,4 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
             this.changeStatus(GlobalStatus.RollbackRetrying);
         }
     }
-
 }

--- a/server/src/main/java/io/seata/server/session/GlobalSession.java
+++ b/server/src/main/java/io/seata/server/session/GlobalSession.java
@@ -19,7 +19,9 @@ import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
@@ -82,6 +84,10 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
 
     private GlobalSessionLock globalSessionLock = new GlobalSessionLock();
 
+    /**
+     * The high reliable session map.
+     */
+    public static Map<GlobalStatus, Set<GlobalSession>> reliableSessionMap = new ConcurrentHashMap<>();
 
     /**
      * Add boolean.
@@ -666,18 +672,24 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
     }
 
     public void asyncCommit() throws TransactionException {
+        // first save global session in reliable map
+        saveInReliableMap(GlobalStatus.AsyncCommitting);
         this.addSessionLifecycleListener(SessionHolder.getAsyncCommittingSessionManager());
         SessionHolder.getAsyncCommittingSessionManager().addGlobalSession(this);
         this.changeStatus(GlobalStatus.AsyncCommitting);
     }
 
     public void queueToRetryCommit() throws TransactionException {
+        // first save global session in reliable map
+        saveInReliableMap(GlobalStatus.CommitRetrying);
         this.addSessionLifecycleListener(SessionHolder.getRetryCommittingSessionManager());
         SessionHolder.getRetryCommittingSessionManager().addGlobalSession(this);
         this.changeStatus(GlobalStatus.CommitRetrying);
     }
 
     public void queueToRetryRollback() throws TransactionException {
+        // first save global session in reliable map
+        saveInReliableMap(GlobalStatus.RollbackRetrying);
         this.addSessionLifecycleListener(SessionHolder.getRetryRollbackingSessionManager());
         SessionHolder.getRetryRollbackingSessionManager().addGlobalSession(this);
         GlobalStatus currentStatus = this.getStatus();
@@ -686,5 +698,18 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
         } else {
             this.changeStatus(GlobalStatus.RollbackRetrying);
         }
+    }
+
+    /**
+     * Save global session in reliable map
+     * @param status global status
+     */
+    private void saveInReliableMap(GlobalStatus status) {
+        Set<GlobalSession> sessions = reliableSessionMap.get(status);
+        if (sessions == null) {
+            sessions = new HashSet<>();
+        }
+        sessions.add(this);
+        reliableSessionMap.put(status, sessions);
     }
 }

--- a/server/src/main/java/io/seata/server/session/GlobalSession.java
+++ b/server/src/main/java/io/seata/server/session/GlobalSession.java
@@ -19,9 +19,7 @@ import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashSet;
-import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
@@ -84,10 +82,6 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
 
     private GlobalSessionLock globalSessionLock = new GlobalSessionLock();
 
-    /**
-     * The high reliable session map.
-     */
-    public static Map<GlobalStatus, Set<GlobalSession>> reliableSessionMap = new ConcurrentHashMap<>();
 
     /**
      * Add boolean.
@@ -672,24 +666,18 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
     }
 
     public void asyncCommit() throws TransactionException {
-        // first save global session in reliable map
-        saveInReliableMap(GlobalStatus.AsyncCommitting);
         this.addSessionLifecycleListener(SessionHolder.getAsyncCommittingSessionManager());
         SessionHolder.getAsyncCommittingSessionManager().addGlobalSession(this);
         this.changeStatus(GlobalStatus.AsyncCommitting);
     }
 
     public void queueToRetryCommit() throws TransactionException {
-        // first save global session in reliable map
-        saveInReliableMap(GlobalStatus.CommitRetrying);
         this.addSessionLifecycleListener(SessionHolder.getRetryCommittingSessionManager());
         SessionHolder.getRetryCommittingSessionManager().addGlobalSession(this);
         this.changeStatus(GlobalStatus.CommitRetrying);
     }
 
     public void queueToRetryRollback() throws TransactionException {
-        // first save global session in reliable map
-        saveInReliableMap(GlobalStatus.RollbackRetrying);
         this.addSessionLifecycleListener(SessionHolder.getRetryRollbackingSessionManager());
         SessionHolder.getRetryRollbackingSessionManager().addGlobalSession(this);
         GlobalStatus currentStatus = this.getStatus();
@@ -700,16 +688,4 @@ public class GlobalSession implements SessionLifecycle, SessionStorable {
         }
     }
 
-    /**
-     * Save global session in reliable map
-     * @param status global status
-     */
-    private void saveInReliableMap(GlobalStatus status) {
-        Set<GlobalSession> sessions = reliableSessionMap.get(status);
-        if (sessions == null) {
-            sessions = new HashSet<>();
-        }
-        sessions.add(this);
-        reliableSessionMap.put(status, sessions);
-    }
 }

--- a/server/src/main/java/io/seata/server/storage/db/session/DataBaseSessionManager.java
+++ b/server/src/main/java/io/seata/server/storage/db/session/DataBaseSessionManager.java
@@ -168,7 +168,7 @@ public class DataBaseSessionManager extends AbstractSessionManager
         if (SessionHolder.ASYNC_COMMITTING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
             return findGlobalSessions(new SessionCondition(GlobalStatus.AsyncCommitting));
         } else if (SessionHolder.RETRY_COMMITTING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
-            return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.CommitRetrying}));
+            return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.CommitRetrying, GlobalStatus.Committing}));
         } else if (SessionHolder.RETRY_ROLLBACKING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
             return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.RollbackRetrying,
                 GlobalStatus.Rollbacking, GlobalStatus.TimeoutRollbacking, GlobalStatus.TimeoutRollbackRetrying}));

--- a/server/src/main/java/io/seata/server/storage/redis/session/RedisSessionManager.java
+++ b/server/src/main/java/io/seata/server/storage/redis/session/RedisSessionManager.java
@@ -166,7 +166,7 @@ public class RedisSessionManager extends AbstractSessionManager
         if (SessionHolder.ASYNC_COMMITTING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
             return findGlobalSessions(new SessionCondition(GlobalStatus.AsyncCommitting));
         } else if (SessionHolder.RETRY_COMMITTING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
-            return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.CommitRetrying}));
+            return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.CommitRetrying, GlobalStatus.Committing}));
         } else if (SessionHolder.RETRY_ROLLBACKING_SESSION_MANAGER_NAME.equalsIgnoreCase(taskName)) {
             return findGlobalSessions(new SessionCondition(new GlobalStatus[] {GlobalStatus.RollbackRetrying,
                 GlobalStatus.Rollbacking, GlobalStatus.TimeoutRollbacking, GlobalStatus.TimeoutRollbackRetrying}));

